OpenEdXMobile/src/main/java/org/edx/mobile/viewModel/InAppPurchasesViewModel.kt

```diff
@@ -110,43 +207,86 @@ class InAppPurchasesViewModel @Inject constructor(
 
                 override fun onError(error: Result.Error) {
                     endLoading()
-                    setError(ErrorMessage.EXECUTE_ORDER_CODE, error.throwable)
+                    dispatchError(
+                        errorType = ErrorMessage.EXECUTE_ORDER_CODE,
+                        errorResId = InAppPurchasesUtils.getErrorMessage(error.throwable),
+                        throwable = error.throwable
+                    )
                 }
             })
     }
 
+    fun dispatchError(
+        errorCode: Int = 0,
+        errorType: Int = 0,
+        httpErrorCode: Int = 0,
+        errorMessage: String? = null,
+        errorResId: Int = 0,
+        throwable: Throwable? = null
+    ) {
+        var actualErrorMessage = errorMessage
+        var actualErrorResId = errorResId
+        throwable?.let {
+            if (TextUtils.isEmpty(errorMessage)) {
+                actualErrorMessage = it.message
+            }
+            if (errorResId == 0) {
+                actualErrorResId = InAppPurchasesUtils.getErrorMessage(it)
+            }
+        }
+        _errorMessage.postValue(
+            ErrorMessage(
+                errorCode = errorType,
+                errorResId = actualErrorResId,
+                throwable = InAppPurchasesException(
+                    errorCode = errorCode,
+                    httpErrorCode = httpErrorCode,
+                    errorMessage = actualErrorMessage
+                )
+            )
+        )
+    }
+
+    /**
+     * To detect and handle courses which are purchased but still not Verified
+     *
+     * @param activity: context of the activity trigger to handle the incomplete purchases flow.
+     * @param userId: handle the incomplete purchases flow for the user.
+     * @param incompletePurchases: A list of pairs of SKU and PurchaseToken of purchased courses
+     * from Payment SDK.
+     *
+     * @return incompletePurchases after dropout the verified one.
+     */
+    fun handleIncompletePurchasesFlow(
```

Comment:
We can re-write this method as, also update the doc
```
fun handleIncompletePurchasesFlow(
        activity: Activity,
        userId: Long,
    ) {
        if (hasIncompletePurchases()) {
            viewModelScope.launch {
                completeCoursePurchaseFlow(activity, userId, incompletePurchases)
            }
            incompletePurchases.drop(1).toMutableList()
        }
    }

Reply:
update the business logic as per the scenario, and update the method accordingly.